services/datamanager/datasync/sync.go

```diff
@@ -124,69 +149,35 @@ func (s *syncer) SyncFile(path string, stopAfter time.Time) {
 	}
 	s.progressLock.Unlock()
 
-	for {
-		if s.cancelCtx.Err() != nil {
-			return
-		}
-
-		if time.Now().After(stopAfter) {
-			return
+	if !s.MarkInProgress(path) {
+		return
+	}
+	defer s.UnmarkInProgress(path)
+	//nolint:gosec
+	f, err := os.Open(path)
+	if err != nil {
+		// Don't log if the file does not exist, because that means it was successfully synced and deleted
+		// in between paths being built and this executing.
+		if !errors.Is(err, os.ErrNotExist) {
+			s.logger.Errorw("error opening file", "error", err)
 		}
+		return
+	}
 
-		select {
-		case <-s.cancelCtx.Done():
-			return
-		// Kick off a sync goroutine if under the limit of goroutines.
-		case s.syncRoutineTracker <- struct{}{}:
-			s.backgroundWorkers.Add(1)
-
-			goutils.PanicCapturingGo(func() {
-				defer s.backgroundWorkers.Done()
-				// At the end, decrement the number of sync routines.
-				defer func() {
-					<-s.syncRoutineTracker
-				}()
-				select {
-				case <-s.cancelCtx.Done():
-					return
-				default:
-					if !s.MarkInProgress(path) {
-						return
-					}
-					defer s.UnmarkInProgress(path)
-					//nolint:gosec
-					f, err := os.Open(path)
-					if err != nil {
-						// Don't log if the file does not exist, because that means it was successfully synced and deleted
-						// in between paths being built and this executing.
-						if !errors.Is(err, os.ErrNotExist) {
-							s.logger.Errorw("error opening file", "error", err)
-						}
-						return
-					}
-
-					if datacapture.IsDataCaptureFile(f) {
-						captureFile, err := datacapture.ReadFile(f)
-						if err != nil {
-							if err = f.Close(); err != nil {
-								s.syncErrs <- errors.Wrap(err, "error closing data capture file")
-							}
-							if err := moveFailedData(f.Name(), s.captureDir); err != nil {
-								s.syncErrs <- errors.Wrap(err, fmt.Sprintf("error moving corrupted data %s", f.Name()))
-							}
-							return
-						}
-						s.syncDataCaptureFile(captureFile)
-					} else {
-						s.syncArbitraryFile(f)
-					}
-				}
-			})
-
+	if datacapture.IsDataCaptureFile(f) {
+		captureFile, err := datacapture.ReadFile(f)
```

Comment:
Just a mental note that this structure I think [is ~4KB of memory](https://cs.opensource.google/go/go/+/refs/tags/go1.22.4:src/bufio/bufio.go;l=19). So 1000 stuck goroutines translates to ~4MB of memory.

I'm not sure if it makes sense for a user of `datacapture.ReadFile` to need a "[write buffer](https://github.com/viamrobotics/rdk/blob/377ebc517678d5935bf19cb3e16c2ff1c360db90/services/datamanager/datacapture/data_capture_file.go#L74)". But just a thought to noodle over some other time. Definitely don't scope creep this PR. 4MB is well within our budget.